src/simulation/ApplyLoad.cpp

```diff
@@ -77,22 +81,104 @@ getUpgradeConfig(Config const& cfg)
+    // and tx count
+    upgradeConfig.maxContractSizeBytes = 65536;
+    upgradeConfig.maxContractDataKeySizeBytes = 250;
+    upgradeConfig.maxContractDataEntrySizeBytes = 65536;
+    upgradeConfig.txMemoryLimit = 41943040;
+    upgradeConfig.liveSorobanStateSizeWindowSampleSize = 30;
+    upgradeConfig.evictionScanSize = 100000;
+    upgradeConfig.startingEvictionScanLevel = 7;
+
+    upgradeConfig.ledgerMaxDiskReadEntries = 100'000'000;
+    upgradeConfig.ledgerMaxDiskReadBytes = 100'000'000;
+    upgradeConfig.ledgerMaxWriteLedgerEntries = 100'000'000;
+    upgradeConfig.ledgerMaxWriteBytes = 100'000'000;
+    upgradeConfig.txMaxDiskReadEntries = 5;
+    upgradeConfig.txMaxFootprintEntries = 5;
+    upgradeConfig.txMaxDiskReadBytes = 10'000;
+    upgradeConfig.txMaxWriteLedgerEntries = 5;
+    upgradeConfig.txMaxWriteBytes = 10'000;
+    upgradeConfig.ledgerMaxTransactionsSizeBytes = 100'000'000;
+    upgradeConfig.txMaxSizeBytes = 10'000;
+    upgradeConfig.txMaxInstructions = 3'000'000;
+    upgradeConfig.txMaxContractEventsSizeBytes = 100'000'000;
+
+    // Increase the default TTL and reduce the rent rate in order to avoid the
+    // state archival and too high rent fees. The apply load test is generally
+    // not concerned about the resource fees.
+    upgradeConfig.minPersistentTTL = 1'000'000'000;
+    upgradeConfig.minTemporaryTTL = 1'000'000'000;
+    upgradeConfig.maxEntryTTL = 1'000'000'001;
+    upgradeConfig.persistentRentRateDenominator = 1'000'000'000'000LL;
+    upgradeConfig.tempRentRateDenominator = 1'000'000'000'000LL;
+
+    // Set the instruction and max tx count just high enough so that we generate
+    // a full ledger. This ensures all available clusters are filled for maximum
+    // parallelism.
+    upgradeConfig.ledgerMaxInstructions = instructionsPerCluster;
+    upgradeConfig.ledgerMaxTxCount = totalTxs;
+
+    releaseAssert(*upgradeConfig.ledgerMaxInstructions > 0);
+    releaseAssert(*upgradeConfig.txMaxInstructions > 0);
+
+    if (*upgradeConfig.ledgerMaxInstructions < *upgradeConfig.txMaxInstructions)
+    {
+        throw std::runtime_error("TPS too low, cannot achieve parallelism");
+    }
+
+    return upgradeConfig;
+}
+}
+
+void
+ApplyLoad::upgradeSettingsForMaxTPS(uint32_t txsToGenerate)
+{
+    // Calculate the actual instructions needed for all transactions. The
+    // ledger max instructions is the total instruction count per cluster. In
+    // order to have max parallelism, we want each cluster to be full, so
+    // upgrade settings such that we have just enough capacity across all
+    // clusters.
+    uint64_t totalInstructions =
+        static_cast<uint64_t>(txsToGenerate) * TxGenerator::SAC_TX_INSTRUCTIONS;
+    uint64_t instructionsPerCluster =
+        totalInstructions /
+        mApp.getConfig().APPLY_LOAD_LEDGER_MAX_DEPENDENT_TX_CLUSTERS;
+
+    // Add a 5% buffer to ensure all transactions can fit
+    instructionsPerCluster =
```

Comment:
This looks a bit suspicious. IIUC the only way to not fit all the txs is if the cluster size is not divisible by the instructions (given that all txs have the same instructions, that is). So wouldn't it be sufficient to just add `SAC_TX_INSTRUCTIONS - 1` to every cluster?